### PR TITLE
move from_spark, etc. to SparkBackend

### PR DESCRIPTION
**Pull request by cseed**
This is part of a larger set of changes to separate the Python front end from Java and Spark to support connecting to a REST API Hail backend.

### CODE REVIEW

File: hail/python/hail/backend/backend.py [hunk 1]
```diff
@@ -43,6 +45,20 @@ def matrix_read_type(self, mir):
         jir = self._to_java_ir(mir)
         return tmatrix._from_java(jir.typ())
 
+    def from_spark(self, df, key):
+        return Table._from_java(Env.hail().table.Table.fromDF(Env.hc()._jhc, df._jdf, key))
+
+    def to_spark(self, t, flatten):
```
Comment (akotlar) [inline hail/python/hail/backend/backend.py]: Since signature of this method has changed (because no longer in class Table), would it be worth adding a type assertion, or comment such as 

// @param t <Table>

Comment (akotlar) [inline hail/python/hail/backend/backend.py]: It seems this is the case, but to check, we don't need `flatten = True` in the signature, because the current calls that need to be flattened by default only come from Table, which retains the `flatten` default?

**Comment by akotlar** [comment]
Exciting, first code-related pull request review! It seems correct. I was wondering how you're testing table.py, backend.py, java.py, and would it be worthwhile to write unit or integration tests for these sections? I'd be happy to work on that if desired.

File: hail/python/hail/backend/backend.py (hunk 1 above)
Comment (cseed) [reply hail/python/hail/backend/backend.py]: In response to your first comment:
 - We use the typecheck decorator to verify argument types on user-facing code.  This has the effect of both documenting the argument  types and verifying their types at runtime.  We don't do this for user facing code.
 - We could use Python3 type hints.  @tpoterba pushed pretty hard on this at on point, and it got quite unwieldy, mostly because of cases where we took advantage of Python type flexibility and it was not easy (or even possible) to statically declare the type.  He can say more.
 - I'm not familiar with your proposed syntax.  In general, I think I'm going to argue against uncheck comments like this, and type hints, if they are unchecked.  I think @danking has been investigating type linters that might analyze type hints, that would be reasonable.
 - In summary, our current internal code has undeclared/undocumented types.  I'm happy to discuss if we should be using type hints, esp. in the context of linters, but that seems outside the scope of this review.

In response to your second comment:
 - I'm not sure I understand what you mean  by "configuring the shape of the table object" and what you're proposing.  Can you outline the code changes you're thinking of?

Comment (cseed) [reply hail/python/hail/backend/backend.py]: Correct!

Comment (akotlar) [reply hail/python/hail/backend/backend.py]: 1. Sure, I wonder if there is a linter like js flow or typescript available in python. At least in the js case there are escape hatches in case the type cannot be statically known (:any) as well as union types (: Table | Other). Will defer that thought.

2. Old version of table.to_spark/to_pandas seemed better encapsulated, in that it handled all necessary transformations on itself (expand_types in particular). 

    In the new version, we have to make sure that `t` implements `expand_types` and returns an object that implements both `_jt`(which is private to Table) and `toDF`, and that may or may not need to call `flatten()`, where something that can be `toDF()` is all we seem to really care about.

    We could reduce this surface by having `backend.to_spark` directly take an object that implements `toDF()` (trait/concept).

```python
# table.py

def to_spark(self, flatten = True):
  t = self.expand_types();
  if flatten:
     t = t.flatten()
  
  return Env.spark_backend('to_spark').to_spark(t._jt)

# backend.py
# Pretending that we have a static/linting-based type checker
def to_spark(self, t : DataFrameTrait):
   return pyspark.sql.DataFrame(t.toDF(Env.hc()._jsql_context), Env.sql_context())
```

This reduces the amount that backend.py needs to know about consumers of its functions, and should have no performance overhead.

However, I'm missing context, and your solution just may be the only sensible one.

Comment (cseed) [reply hail/python/hail/backend/backend.py]: > as well as union types 

There are definitely Python linters.  We had mixed experiences we started using Python 2+ years ago, but they've probably improved.  When we went down the type annotation route (mainly because IDEs use type hints for completion suggestions), we got massive sprawling unions which is part of why we pulled back on this.  But yes, I'm all in favor of linter + managable type annotations.  @tpoterba and @danking will have more thoughts.  Let's make a plan.

No, your suggestion is totally sensible, but I'm still not sure I like it because it splits up the conceptual implementation of to_spark, which feels like it makes the code harder to follow since it is now distributed.   Personally, I'd look at BackEnd.to_spark and wonder, "shit, who's expanding types, is there a bug?"  I'm not sure it matters to the discussion, but BackEnd.to_spark will only ever be called from one place, MatrixTable.to_spark, that is, the backend version IS the implementation of to_spark.  Also, mine has the negligible benefit of getting to an error more quickly if you're using a non-Spark backend.

Comment (akotlar) [reply hail/python/hail/backend/backend.py]: KK, cool!

Yeah, it probably makes more sense to do it your way, indirection can be hard to follow, I see your point, I would find it confusing at some point too, kind of like deep inheritance, which I dislike. Regarding caller, if BackEnd.to_spark is consumed in one place, does it make sense to have the definition a method on the caller (less indirection)?

Comment (cseed) [reply hail/python/hail/backend/backend.py]: > does it make sense to have the definition a method on the caller (less indirection)?

That's the key point of the change: it can't be because it can only be implemented correctly on the Spark backend.  to_spark converts our Table to a Spark dataframe, but if we're not running with Spark (e.g. a local mode or our custom backend), then we can't do it.

Right now Spark is tangled with all the Python code, e.g. SparkBackend is getting the Spark context and Spark SQL context from Env.  Future change will move those objects into the SparkBackend itself.

So to put this in the caller (Table), it would have to cast the Backend to a SparkBackend (there's effectively your indirection) and reach into SparkBack data structures to implement this function.  Therefore, I think it naturally lives on SparkBackend.

If you're now happy with the change now, can you Approve?

Comment (akotlar) [reply hail/python/hail/backend/backend.py]: Thanks for the explanation
